Scripts/Runtime/Jobs/Collections/CollectionAccessController.cs

```diff
@@ -0,0 +1,226 @@
+                    break;
+                case AccessType.SharedRead:
+                    m_State = AcquisitionState.SharedRead;
+                    acquiredHandle = m_SharedReadDependency;
+                    break;
+                default:
+                    throw new ArgumentOutOfRangeException(nameof(accessType), accessType, $"Tried to acquire with {nameof(AccessType)} of {accessType} but no code path satisfies!");
+            }
+
+#if ENABLE_UNITY_COLLECTIONS_CHECKS
+            m_LastHandleAcquired = acquiredHandle;
+#endif
+
+            return acquiredHandle;
+        }
+
+        /// <summary>
+        /// Allows the <see cref="CollectionAccessController{TContext,TKey}"/> to be aware of the work that you are doing
+        /// for a specific <see cref="AccessType"/>. You must call <see cref="Release"/> after any call to <see cref="Acquire"/>
+        /// before you call <see cref="Acquire"/> again.
+        /// </summary>
+        /// <param name="releaseAccessDependency">The <see cref="JobHandle"/> to the job that is doing the reading or
+        /// writing from/to the underlying collection this <see cref="CollectionAccessController{TContext,TKey}"/>
+        /// is gating access to.</param>
+        public void Release(JobHandle releaseAccessDependency)
+        {
+            Debug.Assert(!IsDisposed);
+            ValidateReleaseState(releaseAccessDependency);
+
+            switch (m_State)
+            {
+                case AcquisitionState.Disposing:
+                case AcquisitionState.ExclusiveWrite:
+                    //If you were exclusively writing, then no one else can do any writing or reading until you're done
+                    //In the disposing case, we set to the same dependencies but gate in the checks
+                    m_ExclusiveWriteDependency = JobHandle.CombineDependencies(m_ExclusiveWriteDependency, releaseAccessDependency);
+                    m_SharedWriteDependency = m_ExclusiveWriteDependency;
+                    m_SharedReadDependency = m_ExclusiveWriteDependency;
+                    break;
+                case AcquisitionState.SharedWrite:
+                    //If you were shared writing, then no one else can exclusive write or read until you're done
+                    m_ExclusiveWriteDependency = JobHandle.CombineDependencies(m_ExclusiveWriteDependency, releaseAccessDependency);
+                    m_SharedReadDependency = m_ExclusiveWriteDependency;
+                    break;
+                case AcquisitionState.SharedRead:
+                    //If you were reading, then no one else can do any writing until your reading is done
+                    m_ExclusiveWriteDependency = JobHandle.CombineDependencies(m_ExclusiveWriteDependency, releaseAccessDependency);
+                    m_SharedWriteDependency = m_ExclusiveWriteDependency;
+                    break;
+                case AcquisitionState.Unacquired:
```

Comment:
This and Disposing should probably just be turning into non-debug assertions

Reply:
Done.